pkg/operators/ebpf/stats.go

```diff
@@ -0,0 +1,535 @@
+	ParamProgramStats             = "programs"
+	StatsDSName                   = "bpfstats"
+	EnableBPFStatsParam           = "enable-bpfstats"
+)
+
+func (i *ebpfOperator) GlobalParams() api.Params {
+	return api.Params{
+		{
+			Key:          EnableBPFStatsParam,
+			Description:  "Enable capturing eBPF stats",
+			DefaultValue: "false",
+			TypeHint:     api.TypeBool,
+			Title:        "Enable eBPF stats",
+		},
+	}
+}
+
+func (i *ebpfOperator) Init(params *params.Params) error {
+	if !params.Get(EnableBPFStatsParam).AsBool() {
+		return nil
+	}
+
+	i.gadgetObjs = make(map[operators.GadgetContext]gadgetObjs)
+
+	// Enable stats collection
+	// TODO: when to disable it?
+	if err := bpfstats.EnableBPFStats(); err != nil {
+		return fmt.Errorf("enabling bpf stats: %w", err)
+	}
+
+	return nil
+}
+
+func (i *ebpfOperator) InstantiateDataOperator(
+	gadgetCtx operators.GadgetContext, paramValues api.ParamValues,
+) (operators.DataOperatorInstance, error) {
+	cfg, ok := gadgetCtx.GetVar("config")
+	if !ok {
+		return nil, fmt.Errorf("missing configuration")
+	}
+	v, ok := cfg.(*viper.Viper)
+	if !ok {
+		return nil, fmt.Errorf("invalid configuration format")
+	}
+
+	emitStatsAnn := v.GetBool(fmt.Sprintf("annotations.operator.%s.%s", i.Name(), EmitStatsAnn))
+	if !emitStatsAnn {
+		return nil, nil
+	}
+
+	if i.gadgetObjs == nil {
+		return nil, fmt.Errorf("bpfstats aren't enabled")
+	}
+
+	var err error
+
+	instance := &ebpfOperatorDataInstance{
+		bpfOperator:  i,
+		done:         make(chan struct{}),
+		programStats: paramValues[ParamProgramStats] == "true",
+	}
+
+	instance.ds, err = gadgetCtx.RegisterDataSource(datasource.TypeArray, StatsDSName)
+	if err != nil {
+		return nil, err
+	}
+
+	intervalAnn := paramValues[ParamGadgetStatisticsInterval]
+	if intervalAnn == "" {
+		intervalAnn = "1000ms"
+	}
+	instance.interval, err = time.ParseDuration(intervalAnn)
+	if err != nil {
+		return nil, fmt.Errorf("parsing duration: %w", err)
+	}
+
+	countAnn := paramValues[ParamGadgetStatisticsCount]
+	if countAnn == "" {
+		countAnn = "0"
+	}
+	instance.count, err = strconv.Atoi(countAnn)
+	if err != nil {
+		return nil, fmt.Errorf("parsing count: %w", err)
+	}
+
+	instance.ds.AddAnnotation(api.FetchIntervalAnnotation, intervalAnn)
+	instance.ds.AddAnnotation(api.FetchCountAnnotation, countAnn)
+	instance.ds.AddAnnotation("cli.clear-screen-before", "true")
+
+	if nodeName, ok := os.LookupEnv("NODE_NAME"); ok {
+		instance.nodeNameField, err = instance.ds.AddField("nodeName", api.Kind_String,
+			datasource.WithAnnotations(map[string]string{
+				metadatav1.ColumnsWidthAnnotation: "16",
+				"metrics.type":                    "key",
+			}),
+		)
+		if err != nil {
+			return nil, err
+		}
+		instance.nodeName = nodeName
+	}
+
+	// gadget-specific fields
+	instance.gadgetIDField, err = instance.ds.AddField("gadgetID", api.Kind_String,
+		datasource.WithAnnotations(map[string]string{
+			metadatav1.ColumnsWidthAnnotation: "8",
+			// NOT REALLY SURE
```

Comment:
Why should we need this? The gadgetName is already unique and 1:1 with the ID, so having the ID would just increase the cardinality of the metrics. Or what am I missing?

Reply:
Yes, that's right, the ID shouldn't be a key.